StaticLinkageChecker to support other repositories than Maven central

When I tried to check [the issue of Google Cloud Pubsub with Spring Cloud GCP][1], StaticLinkageChecker was not able to locate the Maven artifact available in https://repo.spring.io/milestone.

https://github.com/shiraji/demo-pubsub/blob/master/build.gradle

```
Exception in thread "main" org.eclipse.aether.resolution.DependencyResolutionException: The following artifacts could not be resolved: org.springframework.cloud:spring-cloud-gcp-starter-pubsub:jar:1.1.0.RC2, org.springframework.cloud:spring-cloud-gcp-starter-trace:jar:1.1.0.RC2: Could not find artifact org.springframework.cloud:spring-cloud-gcp-starter-pubsub:jar:1.1.0.RC2 in central (http://repo1.maven.org/maven2/)
	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:355)
	at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder.resolveCompileTimeDependencies(DependencyGraphBuilder.java:125)
	at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder.getStaticLinkageCheckDependencyGraph(DependencyGraphBuilder.java:158)
	at com.google.cloud.tools.opensource.classpath.ClassPathBuilder.artifactsToDependencyPaths(ClassPathBuilder.java:86)
	at com.google.cloud.tools.opensource.classpath.ClassPathBuilder.artifactsToClasspath(ClassPathBuilder.java:56)
	at com.google.cloud.tools.opensource.classpath.StaticLinkageCheckOption.generateInputClasspath(StaticLinkageCheckOption.java:115)
	at com.google.cloud.tools.opensource.classpath.StaticLinkageChecker.main(StaticLinkageChecker.java:95)
Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: The following artifacts could not be resolved: org.springframework.cloud:spring-cloud-gcp-starter-pubsub:jar:1.1.0.RC2, org.springframework.cloud:spring-cloud-gcp-starter-trace:jar:1.1.0.RC2: Could not find artifact org.springframework.cloud:spring-cloud-gcp-starter-pubsub:jar:1.1.0.RC2 in central (http://repo1.maven.org/maven2/)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:422)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:224)
	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:338)
	... 6 more
Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact org.springframework.cloud:spring-cloud-gcp-starter-pubsub:jar:1.1.0.RC2 in central (http://repo1.maven.org/maven2/)
	at org.eclipse.aether.connector.basic.ArtifactTransportListener.transferFailed(ArtifactTransportListener.java:48)
	at org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run(BasicRepositoryConnector.java:368)
	at org.eclipse.aether.util.concurrency.RunnableErrorForwarder$1.run(RunnableErrorForwarder.java:75)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```


`DependencyGraphBuilder.resolveCompileTimeDependencies` has

```
    collectRequest.addRepository(RepositoryUtility.CENTRAL);
    CollectResult collectResult = system.collectDependencies(session, collectRequest);
    DependencyNode node = collectResult.getRoot();
```


[1]: https://github.com/spring-cloud/spring-cloud-gcp/issues/1373

Even with spring's repository, StaticLinkageChecker throws an exception for log4j-api-java9...

```
Exception in thread "main" org.eclipse.aether.resolution.DependencyResolutionException: The following artifacts could not be resolved: org.apache.logging.log4j:log4j-api-java9:zip:2.11.1, org.apache.logging.log4j:log4j-core-java9:zip:2.11.1: Could not find artifact org.apache.logging.log4j:log4j-api-java9:zip:2.11.1 in central (http://repo1.maven.org/maven2/)
	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:355)
	at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder.resolveCompileTimeDependencies(DependencyGraphBuilder.java:132)
	at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder.getStaticLinkageCheckDependencyGraph(DependencyGraphBuilder.java:165)
	at com.google.cloud.tools.opensource.classpath.ClassPathBuilder.artifactsToDependencyPaths(ClassPathBuilder.java:86)
	at com.google.cloud.tools.opensource.classpath.ClassPathBuilder.artifactsToClasspath(ClassPathBuilder.java:56)
	at com.google.cloud.tools.opensource.classpath.StaticLinkageCheckOption.generateInputClasspath(StaticLinkageCheckOption.java:115)
	at com.google.cloud.tools.opensource.classpath.StaticLinkageChecker.main(StaticLinkageChecker.java:95)
Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: The following artifacts could not be resolved: org.apache.logging.log4j:log4j-api-java9:zip:2.11.1, org.apache.logging.log4j:log4j-core-java9:zip:2.11.1: Could not find artifact org.apache.logging.log4j:log4j-api-java9:zip:2.11.1 in central (http://repo1.maven.org/maven2/)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:422)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:224)
	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies(DefaultRepositorySystem.java:338)
	... 6 more
Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact org.apache.logging.log4j:log4j-api-java9:zip:2.11.1 in central (http://repo1.maven.org/maven2/)
	at org.eclipse.aether.connector.basic.ArtifactTransportListener.transferFailed(ArtifactTransportListener.java:48)
	at org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run(BasicRepositoryConnector.java:368)
	at org.eclipse.aether.util.concurrency.RunnableErrorForwarder$1.run(RunnableErrorForwarder.java:75)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Where is this coming from? (to be investigated)

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
@@ -122,7 +122,7 @@ public class StaticLinkageCheckOption {
 
   static ImmutableList<Path> generateInputClasspath(CommandLine commandLine)
       throws RepositoryException, ParseException {
-    configureMavenRepositories(commandLine);
+    setRepositories(commandLine);
 
     if (commandLine.hasOption("b")) {
       String bomCoordinates = commandLine.getOptionValue("b");
@@ -149,7 +149,7 @@ public class StaticLinkageCheckOption {
     }
   }
 
-  static void configureMavenRepositories(CommandLine commandLine) throws ParseException {
+  static void setRepositories(CommandLine commandLine) throws ParseException {
     if (!commandLine.hasOption("m")) {
       return;
     }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
@@ -19,27 +19,23 @@ package com.google.cloud.tools.opensource.classpath;
 import static com.google.common.collect.ImmutableList.toImmutableList;
 
 import com.google.cloud.tools.opensource.dependencies.RepositoryUtility;
-import com.google.common.base.Splitter;
 import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableSet;
-import com.google.common.collect.Streams;
-import java.net.URI;
-import java.net.URISyntaxException;
-import java.net.URL;
+import com.google.common.collect.Lists;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.util.Arrays;
 import java.util.List;
 import java.util.stream.Stream;
 import org.apache.commons.cli.CommandLine;
 import org.apache.commons.cli.CommandLineParser;
 import org.apache.commons.cli.DefaultParser;
 import org.apache.commons.cli.HelpFormatter;
+import org.apache.commons.cli.Option;
 import org.apache.commons.cli.Options;
 import org.apache.commons.cli.ParseException;
 import org.eclipse.aether.RepositoryException;
 import org.eclipse.aether.artifact.Artifact;
 import org.eclipse.aether.artifact.DefaultArtifact;
-import org.eclipse.aether.repository.RemoteRepository;
 
 /**
  * Option for {@link StaticLinkageChecker}. To construct an input class path, the checker requires
@@ -56,10 +52,6 @@ import org.eclipse.aether.repository.RemoteRepository;
  *     Linkage Checker: Input</a>
  */
 public class StaticLinkageCheckOption {
-
-  // DefaultTransporterProvider.newTransporter checks the transporters for repository URLs
-  private static final ImmutableSet<String> ALLOWED_REPOSITORY_URL_PROTOCOL =
-      ImmutableSet.of("file", "http", "https");
 
   private static final Options options = configureOptions();
 
@@ -89,26 +81,43 @@ public class StaticLinkageCheckOption {
 
   private static Options configureOptions() {
     Options options = new Options();
+
     options.addOption(
         "b", "bom", true, "BOM to generate a class path, specified by its Maven coordinates");
-    options.addOption(
-        "a",
-        "artifacts",
-        true,
-        "Maven coordinates for Maven artifacts (separated by ',') to generate a class path");
-    options.addOption("j", "jars", true, "Jar files (separated by ',') to generate a class path");
+
+    Option artifactOption = Option.builder("a").longOpt("artifacts").hasArgs()
+        .valueSeparator(',')
+        .desc("Maven coordinates for Maven artifacts (separated by ',') to generate a class path")
+        .build();
+    options.addOption(artifactOption);
+
+    Option jarOption = Option.builder("j").longOpt("jars").hasArgs()
+        .valueSeparator(',')
+        .desc("Jar files (separated by ',') to generate a class path")
+        .build();
+    options.addOption(jarOption);
+
     options.addOption(
         "r",
         "report-only-reachable",
         false,
         "To report only linkage errors reachable from entry point");
-    options.addOption(
-        "m",
-        "maven-repository",
-        true,
-        "Maven repository URL to search for dependencies. "
-            + "If this option is used one or more times, the repositories are added "
-            + "in order before the default Maven Central (http://repo1.maven.org/maven2/).");
+
+    Option repositoryOption = Option.builder("m").longOpt("maven-repositories").hasArgs()
+        .valueSeparator(',')
+        .desc("Maven repository URLs to search for dependencies. "
+            + "The repositories are added to a repository list in order before "
+            + "the default Maven Central (http://repo1.maven.org/maven2/).")
+        .build();
+    options.addOption(repositoryOption);
+
+    Option noMavenCentralOption = Option.builder("nm").longOpt("no-maven-central")
+        .hasArg(false)
+        .desc("Used with '-m'. Not to add Maven Central to the repository list. "
+            + "Useful when the tool is used where Maven central is inaccessible")
+        .build();
+    options.addOption(noMavenCentralOption);
+
     return options;
   }
 
@@ -116,26 +125,22 @@ public class StaticLinkageCheckOption {
       throws RepositoryException, ParseException {
     configureMavenRepositories(commandLine);
 
-    Splitter commaSplitter = Splitter.on(",");
-
     if (commandLine.hasOption("b")) {
       String bomCoordinates = commandLine.getOptionValue("b");
       DefaultArtifact bomArtifact = new DefaultArtifact(bomCoordinates);
       List<Artifact> artifactsInBom = RepositoryUtility.readBom(bomArtifact);
       return ClassPathBuilder.artifactsToClasspath(artifactsInBom);
     } else if (commandLine.hasOption("a")) {
-      String mavenCoordinatesOption = commandLine.getOptionValue("a");
+      String[] mavenCoordinatesOption = commandLine.getOptionValues("a");
       ImmutableList<Artifact> artifacts =
-          commaSplitter
-              .splitToList(mavenCoordinatesOption)
-              .stream()
+          Arrays.stream(mavenCoordinatesOption)
               .map(DefaultArtifact::new)
               .collect(toImmutableList());
       return ClassPathBuilder.artifactsToClasspath(artifacts);
     } else if (commandLine.hasOption("j")) {
-      String jarFiles = commandLine.getOptionValue("j");
+      String[] jarFiles = commandLine.getOptionValues("j");
       ImmutableList<Path> jarFilesInArguments =
-          Streams.stream(commaSplitter.split(jarFiles))
+          Arrays.stream(jarFiles)
               .map(name -> Paths.get(name).toAbsolutePath())
               .collect(toImmutableList());
       return jarFilesInArguments;
@@ -149,26 +154,13 @@ public class StaticLinkageCheckOption {
     if (!commandLine.hasOption("m")) {
       return;
     }
-
-    ImmutableList.Builder<RemoteRepository> repositoryListBulder = ImmutableList.builder();
-    for (String mavenRepositoryUrl : commandLine.getOptionValues("m")) {
-      RemoteRepository repository =
-          new RemoteRepository.Builder(mavenRepositoryUrl, "default", mavenRepositoryUrl).build();
-      if (!ALLOWED_REPOSITORY_URL_PROTOCOL.contains(repository.getProtocol())) {
-        helpFormatter.printHelp("StaticLinkageChecker", options);
-        throw new ParseException(
-            "Invalid URL specified for maven repository: " + mavenRepositoryUrl);
-      }
-      try {
-        // Because the protocol is not an empty string, this URI is absolute.
-        new URI(mavenRepositoryUrl);
-      } catch (URISyntaxException ex) {
-        throw new ParseException("Invalid URL syntax for repository: " + mavenRepositoryUrl);
-      }
-
-      repositoryListBulder.add(repository);
+    try {
+      boolean addMavenCentral = !commandLine.hasOption("nm");
+      RepositoryUtility.configureMavenRepositories(Arrays.asList(commandLine.getOptionValues("m")),
+          addMavenCentral);
+    } catch (IllegalArgumentException ex) {
+      throw new ParseException("Invalid URL specified for Maven repositories: "
+          + ex.getMessage());
     }
-    repositoryListBulder.add(RepositoryUtility.CENTRAL);
-    RepositoryUtility.mavenRepositories = repositoryListBulder.build();
   }
 }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
@@ -23,6 +23,9 @@ import com.google.common.base.Splitter;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Streams;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.List;
@@ -155,6 +158,14 @@ public class StaticLinkageCheckOption {
         helpFormatter.printHelp("StaticLinkageChecker", options);
         throw new ParseException(
             "Invalid URL specified for maven repository: " + mavenRepositoryUrl);
+      }
+      try {
+        URI uri = new URI(mavenRepositoryUrl);
+        if (!uri.isAbsolute()) {
+          throw new ParseException("Repository URL must have an absolute path for file");
+        }
+      } catch (URISyntaxException ex) {
+        throw new ParseException("Invalid URL for repository: " + mavenRepositoryUrl);
       }
 
       repositoryListBulder.add(repository);

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
@@ -21,6 +21,7 @@ import static com.google.common.collect.ImmutableList.toImmutableList;
 import com.google.cloud.tools.opensource.dependencies.RepositoryUtility;
 import com.google.common.base.Splitter;
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Streams;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -35,6 +36,7 @@ import org.apache.commons.cli.ParseException;
 import org.eclipse.aether.RepositoryException;
 import org.eclipse.aether.artifact.Artifact;
 import org.eclipse.aether.artifact.DefaultArtifact;
+import org.eclipse.aether.repository.RemoteRepository;
 
 /**
  * Option for {@link StaticLinkageChecker}. To construct an input class path, the checker requires
@@ -51,13 +53,17 @@ import org.eclipse.aether.artifact.DefaultArtifact;
  *     Linkage Checker: Input</a>
  */
 public class StaticLinkageCheckOption {
-  
+
+  // DefaultTransporterProvider.newTransporter checks the transporters for repository URLs
+  private static final ImmutableSet<String> ALLOWED_REPOSITORY_URL_PROTOCOL =
+      ImmutableSet.of("file", "http", "https");
+
   private static final Options options = configureOptions();
 
   private static final HelpFormatter helpFormatter = new HelpFormatter();
 
   static CommandLine readCommandLine(String[] arguments) throws ParseException {
-    // TODO is this reentrant? Can we reuse it? 
+    // TODO is this reentrant? Can we reuse it?
     // https://issues.apache.org/jira/browse/CLI-291
     CommandLineParser parser = new DefaultParser();
 
@@ -95,9 +101,11 @@ public class StaticLinkageCheckOption {
         "To report only linkage errors reachable from entry point");
     options.addOption(
         "m",
-        "maven-repositories",
+        "maven-repository",
         true,
-        "Maven repositories (separated by ',') to search dependencies");
+        "Maven repository URL to search for dependencies. "
+            + "If this option is used one or more times, the repositories are added "
+            + "in order before the default Maven Central (http://repo1.maven.org/maven2/).");
     return options;
   }
 
@@ -134,13 +142,24 @@ public class StaticLinkageCheckOption {
     }
   }
 
-  static void configureMavenRepositories(CommandLine commandLine) {
+  static void configureMavenRepositories(CommandLine commandLine) throws ParseException {
     if (!commandLine.hasOption("m")) {
       return;
     }
-    Splitter commaSplitter = Splitter.on(",");
-    for (String mavenRepositoryUrl : commaSplitter.splitToList(commandLine.getOptionValue("m"))) {
-      RepositoryUtility.registerMavenRepository(mavenRepositoryUrl);
+
+    ImmutableList.Builder<RemoteRepository> repositoryListBulder = ImmutableList.builder();
+    for (String mavenRepositoryUrl : commandLine.getOptionValues("m")) {
+      RemoteRepository repository =
+          new RemoteRepository.Builder(mavenRepositoryUrl, "default", mavenRepositoryUrl).build();
+      if (!ALLOWED_REPOSITORY_URL_PROTOCOL.contains(repository.getProtocol())) {
+        helpFormatter.printHelp("StaticLinkageChecker", options);
+        throw new ParseException(
+            "Invalid URL specified for maven repository: " + mavenRepositoryUrl);
+      }
+
+      repositoryListBulder.add(repository);
     }
+    repositoryListBulder.add(RepositoryUtility.CENTRAL);
+    RepositoryUtility.mavenRepositories = repositoryListBulder.build();
   }
 }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
@@ -56,7 +56,7 @@ public class StaticLinkageCheckOption {
 
   private static final HelpFormatter helpFormatter = new HelpFormatter();
 
-  static CommandLine readCommandLine(String[] arguments) throws ParseException {
+  static CommandLine readCommandLine(String... arguments) throws ParseException {
     // TODO is this reentrant? Can we reuse it?
     // https://issues.apache.org/jira/browse/CLI-291
     CommandLineParser parser = new DefaultParser();
@@ -110,11 +110,14 @@ public class StaticLinkageCheckOption {
         .build();
     options.addOption(repositoryOption);
 
-    Option noMavenCentralOption = Option.builder("nm").longOpt("no-maven-central")
-        .hasArg(false)
-        .desc("Used with '-m'. Not to add Maven Central to the repository list. "
-            + "Useful when the tool is used where Maven central is inaccessible")
-        .build();
+    Option noMavenCentralOption =
+        Option.builder("nm")
+            .longOpt("no-maven-central")
+            .hasArg(false)
+            .desc(
+                "Do not search Maven Central in addition to the repositories specified by -m. "
+                    + "Useful when Maven Central is inaccessible.")
+            .build();
     options.addOption(noMavenCentralOption);
 
     return options;

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
@@ -20,7 +20,6 @@ import static com.google.common.collect.ImmutableList.toImmutableList;
 
 import com.google.cloud.tools.opensource.dependencies.RepositoryUtility;
 import com.google.common.collect.ImmutableList;
-import com.google.common.collect.Lists;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.Arrays;
@@ -156,7 +155,7 @@ public class StaticLinkageCheckOption {
     }
     try {
       boolean addMavenCentral = !commandLine.hasOption("nm");
-      RepositoryUtility.configureMavenRepositories(Arrays.asList(commandLine.getOptionValues("m")),
+      RepositoryUtility.setRepositories(Arrays.asList(commandLine.getOptionValues("m")),
           addMavenCentral);
     } catch (IllegalArgumentException ex) {
       throw new ParseException("Invalid URL specified for Maven repositories: "

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
@@ -93,11 +93,18 @@ public class StaticLinkageCheckOption {
         "report-only-reachable",
         false,
         "To report only linkage errors reachable from entry point");
+    options.addOption(
+        "m",
+        "maven-repositories",
+        true,
+        "Maven repositories (separated by ',') to search dependencies");
     return options;
   }
 
   static ImmutableList<Path> generateInputClasspath(CommandLine commandLine)
       throws RepositoryException, ParseException {
+    configureMavenRepositories(commandLine);
+
     Splitter commaSplitter = Splitter.on(",");
 
     if (commandLine.hasOption("b")) {
@@ -124,6 +131,16 @@ public class StaticLinkageCheckOption {
     } else {
       helpFormatter.printHelp("StaticLinkageChecker", options);
       throw new ParseException("Missing argument");
+    }
+  }
+
+  static void configureMavenRepositories(CommandLine commandLine) {
+    if (!commandLine.hasOption("m")) {
+      return;
+    }
+    Splitter commaSplitter = Splitter.on(",");
+    for (String mavenRepositoryUrl : commaSplitter.splitToList(commandLine.getOptionValue("m"))) {
+      RepositoryUtility.registerMavenRepository(mavenRepositoryUrl);
     }
   }
 }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOption.java
@@ -160,12 +160,10 @@ public class StaticLinkageCheckOption {
             "Invalid URL specified for maven repository: " + mavenRepositoryUrl);
       }
       try {
-        URI uri = new URI(mavenRepositoryUrl);
-        if (!uri.isAbsolute()) {
-          throw new ParseException("Repository URL must have an absolute path for file");
-        }
+        // Because the protocol is not an empty string, this URI is absolute.
+        new URI(mavenRepositoryUrl);
       } catch (URISyntaxException ex) {
-        throw new ParseException("Invalid URL for repository: " + mavenRepositoryUrl);
+        throw new ParseException("Invalid URL syntax for repository: " + mavenRepositoryUrl);
       }
 
       repositoryListBulder.add(repository);

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java
@@ -63,7 +63,7 @@ public class DependencyGraphBuilder {
     } else {
       // Since we only load the dependency graph, not actually use the
       // dependency, it doesn't matter a great deal which one we pick.
-      System.setProperty("os.detected.classifier", "linux-x86_64");      
+      System.setProperty("os.detected.classifier", "linux-x86_64");
     }
   }
 

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java
@@ -62,7 +62,7 @@ public class DependencyGraphBuilder {
       System.setProperty("os.detected.classifier", "linux-x86_64");
     } else {
       // Since we only load the dependency graph, not actually use the
-      // dependency, it doesn't matter a great deal which one we pick.
+      // dependency, it doesn't matter a great deal which one we pick.      
       System.setProperty("os.detected.classifier", "linux-x86_64");
     }
   }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java
@@ -62,8 +62,8 @@ public class DependencyGraphBuilder {
       System.setProperty("os.detected.classifier", "linux-x86_64");
     } else {
       // Since we only load the dependency graph, not actually use the
-      // dependency, it doesn't matter a great deal which one we pick.      
-      System.setProperty("os.detected.classifier", "linux-x86_64");
+      // dependency, it doesn't matter a great deal which one we pick.
+      System.setProperty("os.detected.classifier", "linux-x86_64");      
     }
   }
 

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java
@@ -50,7 +50,7 @@ public class DependencyGraphBuilder {
   private static final Logger logger = Logger.getLogger(DependencyGraphBuilder.class.getName());
   
   private static final RepositorySystem system = RepositoryUtility.newRepositorySystem();
-  
+
   static {
     // os.detected.classifier system property used to select Netty deps
     String OS = System.getProperty("os.name", "generic").toLowerCase(Locale.ENGLISH);
@@ -61,9 +61,9 @@ public class DependencyGraphBuilder {
     } else if (OS.indexOf("nux") >= 0) {
       System.setProperty("os.detected.classifier", "linux-x86_64");
     } else {
-      // Since we only load the dependency graph, not actually use the 
+      // Since we only load the dependency graph, not actually use the
       // dependency, it doesn't matter a great deal which one we pick.
-      System.setProperty("os.detected.classifier", "linux-x86_64");      
+      System.setProperty("os.detected.classifier", "linux-x86_64");
     }
   }
 
@@ -113,7 +113,7 @@ public class DependencyGraphBuilder {
     } else {
       collectRequest.setDependencies(dependencyList);
     }
-    collectRequest.addRepository(RepositoryUtility.CENTRAL);
+    RepositoryUtility.addRepositoriesToRequest(collectRequest);
     CollectResult collectResult = system.collectDependencies(session, collectRequest);
     DependencyNode node = collectResult.getRoot();
 

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
@@ -17,7 +17,10 @@
 package com.google.cloud.tools.opensource.dependencies;
 
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
 import java.io.File;
+import java.net.URI;
+import java.net.URISyntaxException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -25,6 +28,7 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 import java.util.logging.Logger;
+import org.apache.commons.cli.ParseException;
 import org.apache.maven.repository.internal.MavenRepositorySystemUtils;
 import org.eclipse.aether.DefaultRepositorySystemSession;
 import org.eclipse.aether.RepositorySystem;
@@ -61,7 +65,11 @@ public final class RepositoryUtility {
   public static final RemoteRepository CENTRAL =
       new RemoteRepository.Builder("central", "default", "http://repo1.maven.org/maven2/").build();
 
-  public static ImmutableList<RemoteRepository> mavenRepositories = ImmutableList.of(CENTRAL);
+  private static ImmutableList<RemoteRepository> mavenRepositories = ImmutableList.of(CENTRAL);
+
+  // DefaultTransporterProvider.newTransporter checks these transporters
+  private static final ImmutableSet<String> ALLOWED_REPOSITORY_URL_SCHEMES =
+      ImmutableSet.of("file", "http", "https");
 
   private RepositoryUtility() {}
 
@@ -203,6 +211,39 @@ public final class RepositoryUtility {
       }
     }
     return managedDependencies;
+  }
+
+  /**
+   * Configures Maven repositories to search for dependencies.
+   *
+   * @param addMavenCentral if true, add Maven Central to the end of the repository list
+   */
+  public static void configureMavenRepositories(Iterable<String> mavenRepositoryUrls,
+      boolean addMavenCentral) {
+    ImmutableList.Builder<RemoteRepository> repositoryListBuilder = ImmutableList.builder();
+    for (String mavenRepositoryUrl : mavenRepositoryUrls) {
+      RemoteRepository repository =
+          new RemoteRepository.Builder(null, "default", mavenRepositoryUrl).build();
+      if (!ALLOWED_REPOSITORY_URL_SCHEMES.contains(repository.getProtocol())) {
+        throw new IllegalArgumentException(
+            "Protocol: " + repository.getProtocol() + " is not in "
+                + ALLOWED_REPOSITORY_URL_SCHEMES);
+      }
+      try {
+        // Because the protocol is not an empty string, this URI is absolute.
+        new URI(mavenRepositoryUrl);
+      } catch (URISyntaxException ex) {
+        throw new IllegalArgumentException(
+            "Invalid URL syntax: " + mavenRepositoryUrl);
+      }
+      repositoryListBuilder.add(repository);
+    }
+
+    if (addMavenCentral) {
+      repositoryListBuilder.add(CENTRAL);
+    }
+
+    mavenRepositories = repositoryListBuilder.build();
   }
 
   static void addRepositoriesToRequest(CollectRequest collectRequest) {

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
@@ -16,6 +16,8 @@
 
 package com.google.cloud.tools.opensource.dependencies;
 
+import static com.google.common.base.Preconditions.checkArgument;
+
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import java.io.File;
@@ -216,7 +218,7 @@ public final class RepositoryUtility {
    * Sets {@link #mavenRepositories} to search for dependencies.
    *
    * @param addMavenCentral if true, add Maven Central to the end of the repository list
-   * @throws IllegalArgumentException if a URL is malformed or not having an allowed schema
+   * @throws IllegalArgumentException if a URL is malformed or not having an allowed scheme
    */
   public static void setRepositories(
       Iterable<String> mavenRepositoryUrls, boolean addMavenCentral) {
@@ -231,13 +233,13 @@ public final class RepositoryUtility {
 
       RemoteRepository repository =
           new RemoteRepository.Builder(null, "default", mavenRepositoryUrl).build();
-      if (!ALLOWED_REPOSITORY_URL_SCHEMES.contains(repository.getProtocol())) {
-        throw new IllegalArgumentException(
-            "Schema: '"
-                + repository.getProtocol()
-                + "' is not in "
-                + ALLOWED_REPOSITORY_URL_SCHEMES);
-      }
+
+      checkArgument(
+          ALLOWED_REPOSITORY_URL_SCHEMES.contains(repository.getProtocol()),
+          "Scheme: '%s' is not in %s",
+          repository.getProtocol(),
+          ALLOWED_REPOSITORY_URL_SCHEMES);
+
       repositoryListBuilder.add(repository);
     }
 

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
@@ -217,24 +217,25 @@ public final class RepositoryUtility {
    * Configures Maven repositories to search for dependencies.
    *
    * @param addMavenCentral if true, add Maven Central to the end of the repository list
+   * @throws IllegalArgumentException if a URL is malformed or not having an allowed schema
    */
-  public static void configureMavenRepositories(Iterable<String> mavenRepositoryUrls,
-      boolean addMavenCentral) {
+  public static void setRepositories(
+      Iterable<String> mavenRepositoryUrls, boolean addMavenCentral) {
     ImmutableList.Builder<RemoteRepository> repositoryListBuilder = ImmutableList.builder();
     for (String mavenRepositoryUrl : mavenRepositoryUrls) {
+      try {
+        // Because the protocol is not an empty string, this URI is absolute.
+        new URI(mavenRepositoryUrl);
+      } catch (URISyntaxException ex) {
+        throw new IllegalArgumentException("Invalid URL syntax: " + mavenRepositoryUrl);
+      }
+
       RemoteRepository repository =
           new RemoteRepository.Builder(null, "default", mavenRepositoryUrl).build();
       if (!ALLOWED_REPOSITORY_URL_SCHEMES.contains(repository.getProtocol())) {
         throw new IllegalArgumentException(
             "Protocol: " + repository.getProtocol() + " is not in "
                 + ALLOWED_REPOSITORY_URL_SCHEMES);
-      }
-      try {
-        // Because the protocol is not an empty string, this URI is absolute.
-        new URI(mavenRepositoryUrl);
-      } catch (URISyntaxException ex) {
-        throw new IllegalArgumentException(
-            "Invalid URL syntax: " + mavenRepositoryUrl);
       }
       repositoryListBuilder.add(repository);
     }
@@ -246,7 +247,8 @@ public final class RepositoryUtility {
     mavenRepositories = repositoryListBuilder.build();
   }
 
-  static void addRepositoriesToRequest(CollectRequest collectRequest) {
+  /** Adds Maven repositories to {@code collectRequest}. */
+  public static void addRepositoriesToRequest(CollectRequest collectRequest) {
     for (RemoteRepository repository : mavenRepositories) {
       collectRequest.addRepository(repository);
     }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
@@ -16,6 +16,7 @@
 
 package com.google.cloud.tools.opensource.dependencies;
 
+import com.google.common.collect.Lists;
 import java.io.File;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -30,6 +31,7 @@ import org.eclipse.aether.RepositorySystem;
 import org.eclipse.aether.RepositorySystemSession;
 import org.eclipse.aether.artifact.Artifact;
 import org.eclipse.aether.artifact.ArtifactProperties;
+import org.eclipse.aether.collection.CollectRequest;
 import org.eclipse.aether.collection.DependencyCollectionContext;
 import org.eclipse.aether.collection.DependencySelector;
 import org.eclipse.aether.connector.basic.BasicRepositoryConnectorFactory;
@@ -58,6 +60,8 @@ public final class RepositoryUtility {
   
   public static final RemoteRepository CENTRAL =
       new RemoteRepository.Builder("central", "default", "http://repo1.maven.org/maven2/").build();
+
+  private static final List<RemoteRepository> mavenRepositories = Lists.newArrayList(CENTRAL);
 
   private RepositoryUtility() {}
 
@@ -161,7 +165,10 @@ public final class RepositoryUtility {
     RepositorySystemSession session = RepositoryUtility.newSession(system);
 
     ArtifactDescriptorRequest request = new ArtifactDescriptorRequest();
-    request.addRepository(RepositoryUtility.CENTRAL);
+
+    for (RemoteRepository repository : mavenRepositories) {
+      request.addRepository(repository);
+    }
     request.setArtifact(artifact);
 
     ArtifactDescriptorResult resolved = system.readArtifactDescriptor(session, request);
@@ -198,4 +205,14 @@ public final class RepositoryUtility {
     return managedDependencies;
   }
 
+  public static void registerMavenRepository(String mavenRepositoryUrl) {
+    mavenRepositories.add(
+        new RemoteRepository.Builder(mavenRepositoryUrl, "default", mavenRepositoryUrl).build());
+  }
+
+  static void addRepositoriesToRequest(CollectRequest collectRequest) {
+    for (RemoteRepository repository : mavenRepositories) {
+      collectRequest.addRepository(repository);
+    }
+  }
 }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
@@ -28,7 +28,6 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 import java.util.logging.Logger;
-import org.apache.commons.cli.ParseException;
 import org.apache.maven.repository.internal.MavenRepositorySystemUtils;
 import org.eclipse.aether.DefaultRepositorySystemSession;
 import org.eclipse.aether.RepositorySystem;
@@ -214,7 +213,7 @@ public final class RepositoryUtility {
   }
 
   /**
-   * Configures Maven repositories to search for dependencies.
+   * Sets {@link #mavenRepositories} to search for dependencies.
    *
    * @param addMavenCentral if true, add Maven Central to the end of the repository list
    * @throws IllegalArgumentException if a URL is malformed or not having an allowed schema
@@ -234,7 +233,9 @@ public final class RepositoryUtility {
           new RemoteRepository.Builder(null, "default", mavenRepositoryUrl).build();
       if (!ALLOWED_REPOSITORY_URL_SCHEMES.contains(repository.getProtocol())) {
         throw new IllegalArgumentException(
-            "Protocol: " + repository.getProtocol() + " is not in "
+            "Schema: '"
+                + repository.getProtocol()
+                + "' is not in "
                 + ALLOWED_REPOSITORY_URL_SCHEMES);
       }
       repositoryListBuilder.add(repository);
@@ -247,7 +248,7 @@ public final class RepositoryUtility {
     mavenRepositories = repositoryListBuilder.build();
   }
 
-  /** Adds Maven repositories to {@code collectRequest}. */
+  /** Adds {@link #mavenRepositories} to {@code collectRequest}. */
   public static void addRepositoriesToRequest(CollectRequest collectRequest) {
     for (RemoteRepository repository : mavenRepositories) {
       collectRequest.addRepository(repository);

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/RepositoryUtility.java
@@ -16,7 +16,7 @@
 
 package com.google.cloud.tools.opensource.dependencies;
 
-import com.google.common.collect.Lists;
+import com.google.common.collect.ImmutableList;
 import java.io.File;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -61,7 +61,7 @@ public final class RepositoryUtility {
   public static final RemoteRepository CENTRAL =
       new RemoteRepository.Builder("central", "default", "http://repo1.maven.org/maven2/").build();
 
-  private static final List<RemoteRepository> mavenRepositories = Lists.newArrayList(CENTRAL);
+  public static ImmutableList<RemoteRepository> mavenRepositories = ImmutableList.of(CENTRAL);
 
   private RepositoryUtility() {}
 
@@ -203,11 +203,6 @@ public final class RepositoryUtility {
       }
     }
     return managedDependencies;
-  }
-
-  public static void registerMavenRepository(String mavenRepositoryUrl) {
-    mavenRepositories.add(
-        new RemoteRepository.Builder(mavenRepositoryUrl, "default", mavenRepositoryUrl).build());
   }
 
   static void addRepositoriesToRequest(CollectRequest collectRequest) {

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
@@ -16,8 +16,15 @@
 
 package com.google.cloud.tools.opensource.classpath;
 
+import com.google.common.collect.ImmutableList;
+import com.google.common.truth.Truth;
+import java.nio.file.Path;
+import java.util.List;
 import org.apache.commons.cli.CommandLine;
 import org.apache.commons.cli.ParseException;
+import org.eclipse.aether.RepositoryException;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.artifact.DefaultArtifact;
 import org.junit.Assert;
 import org.junit.Test;
 
@@ -57,5 +64,21 @@ public class StaticLinkageCheckOptionTest {
     } catch (ParseException ex) {
       Assert.assertEquals("Unrecognized option: -x", ex.getMessage());
     }
+  }
+
+  @Test
+  public void testConfigureAdditionalMavenRepositories_addingSpringRepository()
+      throws ParseException, RepositoryException {
+    CommandLine commandLine =
+        StaticLinkageCheckOption.readCommandLine(
+            new String[] {"-m", "https://repo.spring.io/milestone"});
+    StaticLinkageCheckOption.configureMavenRepositories(commandLine);
+
+    // This artifact does not exist in Maven central, but it is in Spring's repository
+    // Spring-asm is used here because it does not have complex dependencies
+    Artifact artifact = new DefaultArtifact("org.springframework:spring-asm:3.1.0.RC2");
+
+    List<Path> paths = ClassPathBuilder.artifactsToClasspath(ImmutableList.of(artifact));
+    Truth.assertThat(paths).isNotEmpty();
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
@@ -125,7 +125,7 @@ public class StaticLinkageCheckOptionTest {
 
     List<RemoteRepository> actualRepositories = collectRequest.getRepositories();
     Truth.assertThat(actualRepositories).hasSize(1);
-    Truth.assertThat(actualRepositories).isEqualTo("repo.spring.io");
+    Truth.assertThat(actualRepositories.get(0).getHost()).isEqualTo("repo.spring.io");
   }
 
   @Test

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
@@ -102,7 +102,7 @@ public class StaticLinkageCheckOptionTest {
     CommandLine commandLine =
         StaticLinkageCheckOption.readCommandLine(
             new String[] {"-m", "https://repo.spring.io/milestone"});
-    StaticLinkageCheckOption.configureMavenRepositories(commandLine);
+    StaticLinkageCheckOption.setRepositories(commandLine);
 
     // This artifact does not exist in Maven central, but it is in Spring's repository
     // Spring-asm is used here because it does not have complex dependencies
@@ -118,7 +118,7 @@ public class StaticLinkageCheckOptionTest {
     CommandLine commandLine =
         StaticLinkageCheckOption.readCommandLine(
             new String[] {"-m", "https://repo.spring.io/milestone", "--no-maven-central"});
-    StaticLinkageCheckOption.configureMavenRepositories(commandLine);
+    StaticLinkageCheckOption.setRepositories(commandLine);
 
     CollectRequest collectRequest = new CollectRequest();
     RepositoryUtility.addRepositoriesToRequest(collectRequest);
@@ -142,7 +142,7 @@ public class StaticLinkageCheckOptionTest {
     CommandLine commandLine =
         StaticLinkageCheckOption.readCommandLine(new String[]{"-m", repositoryUrl});
     try {
-      StaticLinkageCheckOption.configureMavenRepositories(commandLine);
+      StaticLinkageCheckOption.setRepositories(commandLine);
       Assert.fail("URL " + repositoryUrl + " should be invalidated");
     } catch (ParseException ex) {
       // pass
@@ -155,7 +155,7 @@ public class StaticLinkageCheckOptionTest {
         StaticLinkageCheckOption.readCommandLine(new String[] {"-m", "file:///var/tmp"});
 
     // This method should not raise an exception
-    StaticLinkageCheckOption.configureMavenRepositories(commandLine);
+    StaticLinkageCheckOption.setRepositories(commandLine);
   }
 
   @Test

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
@@ -35,7 +35,7 @@ public class StaticLinkageCheckOptionTest {
   @After
   public void cleanup() {
     // Resets the effect of configureMavenRepositories
-    RepositoryUtility.mavenRepositories = ImmutableList.of(RepositoryUtility.CENTRAL);
+    RepositoryUtility.configureMavenRepositories(ImmutableList.of(), true);
   }
 
   @Test
@@ -62,6 +62,26 @@ public class StaticLinkageCheckOptionTest {
           ex.getMessage());
     }
   }
+
+  @Test
+  public void testReadCommandLine_multipleArtifacts() throws ParseException {
+    String[] arguments = {
+        "--artifacts", "abc.com:abc:1.1,abc.com:abc-util:1.2",
+        "--report-only-reachable"
+    };
+    CommandLine commandLine = StaticLinkageCheckOption.readCommandLine(arguments);
+    Truth.assertThat(commandLine.getOptionValues("a")).hasLength(2);
+  }
+
+  @Test
+  public void testReadCommandLine_multipleJars() throws ParseException {
+    String[] arguments = {
+        "-j", "/foo/bar/A.jar,/foo/bar/B.jar,/foo/bar/C.jar",
+    };
+    CommandLine commandLine = StaticLinkageCheckOption.readCommandLine(arguments);
+    Truth.assertThat(commandLine.getOptionValues("j")).hasLength(3);
+  }
+
 
   @Test
   public void parseCommandLineOptions_invalidOption() {
@@ -91,21 +111,41 @@ public class StaticLinkageCheckOptionTest {
   }
 
   @Test
+  public void testConfigureAdditionalMavenRepositories_notToUseMavenCentral()
+      throws ParseException {
+    CommandLine commandLine =
+        StaticLinkageCheckOption.readCommandLine(
+            new String[] {"-m", "https://repo.spring.io/milestone", "--no-maven-central"});
+    StaticLinkageCheckOption.configureMavenRepositories(commandLine);
+
+    Artifact artifact = new DefaultArtifact("io.grpc:grpc-auth:1.15.1");
+
+    try {
+      ClassPathBuilder.artifactsToClasspath(ImmutableList.of(artifact));
+      Assert.fail("gRPC artifact should not have been resolved without Maven Central");
+    } catch (RepositoryException ex) {
+      // pass
+    }
+  }
+
+  @Test
   public void testConfigureAdditionalMavenRepositories_invalidRepositoryUrl()
       throws ParseException {
+    assertMavenRepositoryIsInvalid("foobar");
+    assertMavenRepositoryIsInvalid("_http_file__https");
+    assertMavenRepositoryIsInvalid("localhost/abc");
+    assertMavenRepositoryIsInvalid("http://foo^bar");
+  }
 
-    List<String> invalidUrls =
-        ImmutableList.of("foobar", "_http_file__https", "localhost/abc", "http://foo^bar");
-
-    for (String invalidUrl : invalidUrls) {
-      CommandLine commandLine =
-          StaticLinkageCheckOption.readCommandLine(new String[] {"-m", invalidUrl});
-      try {
-        StaticLinkageCheckOption.configureMavenRepositories(commandLine);
-        Assert.fail("URL " + invalidUrl + " should be invalidated");
-      } catch (ParseException ex) {
-        // pass
-      }
+  private static void assertMavenRepositoryIsInvalid(String repositoryUrl)
+      throws ParseException {
+    CommandLine commandLine =
+        StaticLinkageCheckOption.readCommandLine(new String[]{"-m", repositoryUrl});
+    try {
+      StaticLinkageCheckOption.configureMavenRepositories(commandLine);
+      Assert.fail("URL " + repositoryUrl + " should be invalidated");
+    } catch (ParseException ex) {
+      // pass
     }
   }
 
@@ -116,5 +156,13 @@ public class StaticLinkageCheckOptionTest {
 
     // This method should not raise an exception
     StaticLinkageCheckOption.configureMavenRepositories(commandLine);
+  }
+
+  @Test
+  public void testReadCommandLine_multipleRepositoriesSeparatedByComma() throws ParseException {
+    CommandLine commandLine =
+        StaticLinkageCheckOption.readCommandLine(
+            new String[]{"-m", "file:///var/tmp,https://repo.spring.io/milestone"});
+    Truth.assertThat(commandLine.getOptionValues("m")).hasLength(2);
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
@@ -42,21 +42,21 @@ public class StaticLinkageCheckOptionTest {
 
   @Test
   public void parseCommandLineOptions_shortOptions_bom() throws ParseException {
-    String[] arguments = {"-b", "abc.com:dummy:1.2", "-r"};
-    CommandLine parsedOption = StaticLinkageCheckOption.readCommandLine(arguments);
+    CommandLine parsedOption =
+        StaticLinkageCheckOption.readCommandLine("-b", "abc.com:dummy:1.2", "-r");
 
     Assert.assertEquals("abc.com:dummy:1.2", parsedOption.getOptionObject('b'));
   }
 
   @Test
   public void parseCommandLineOptions_duplicates() {
-    String[] arguments = {
-        "--artifacts", "abc.com:abc:1.1,abc.com:abc-util:1.2",
-        "-b", "abc.com:dummy:1.2",
-        "--report-only-reachable"
-    };
     try {
-      StaticLinkageCheckOption.readCommandLine(arguments);
+      StaticLinkageCheckOption.readCommandLine(
+          "--artifacts",
+          "abc.com:abc:1.1,abc.com:abc-util:1.2",
+          "-b",
+          "abc.com:dummy:1.2",
+          "--report-only-reachable");
       Assert.fail();
     } catch (ParseException ex) {
       Assert.assertEquals(
@@ -67,29 +67,25 @@ public class StaticLinkageCheckOptionTest {
 
   @Test
   public void testReadCommandLine_multipleArtifacts() throws ParseException {
-    String[] arguments = {
-        "--artifacts", "abc.com:abc:1.1,abc.com:abc-util:1.2",
-        "--report-only-reachable"
-    };
-    CommandLine commandLine = StaticLinkageCheckOption.readCommandLine(arguments);
+    CommandLine commandLine =
+        StaticLinkageCheckOption.readCommandLine(
+            "--artifacts", "abc.com:abc:1.1,abc.com:abc-util:1.2", "--report-only-reachable");
     Truth.assertThat(commandLine.getOptionValues("a")).hasLength(2);
   }
 
   @Test
   public void testReadCommandLine_multipleJars() throws ParseException {
-    String[] arguments = {
-        "-j", "/foo/bar/A.jar,/foo/bar/B.jar,/foo/bar/C.jar",
-    };
-    CommandLine commandLine = StaticLinkageCheckOption.readCommandLine(arguments);
+    CommandLine commandLine =
+        StaticLinkageCheckOption.readCommandLine(
+            "-j", "/foo/bar/A.jar,/foo/bar/B.jar,/foo/bar/C.jar");
     Truth.assertThat(commandLine.getOptionValues("j")).hasLength(3);
   }
 
 
   @Test
   public void parseCommandLineOptions_invalidOption() {
-    String[] arguments = {"-x"}; // No such option
     try {
-      StaticLinkageCheckOption.readCommandLine(arguments);
+      StaticLinkageCheckOption.readCommandLine("-x"); // No such option
       Assert.fail();
     } catch (ParseException ex) {
       Assert.assertEquals("Unrecognized option: -x", ex.getMessage());

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
@@ -93,17 +93,19 @@ public class StaticLinkageCheckOptionTest {
   @Test
   public void testConfigureAdditionalMavenRepositories_invalidRepositoryUrl()
       throws ParseException {
-    CommandLine commandLine =
-        StaticLinkageCheckOption.readCommandLine(
-            new String[] {"-m", "https://repo.spring.io/milestone", "-m", "foobar"});
 
-    try {
-      StaticLinkageCheckOption.configureMavenRepositories(commandLine);
-      Assert.fail();
-    } catch (ParseException ex) {
-      // pass
-      Truth.assertThat(ex.getMessage())
-          .isEqualTo("Invalid URL specified for maven repository: foobar");
+    List<String> invalidUrls =
+        ImmutableList.of("foobar", "_http_file__https", "localhost/abc", "http://foo^bar");
+
+    for (String invalidUrl : invalidUrls) {
+      CommandLine commandLine =
+          StaticLinkageCheckOption.readCommandLine(new String[] {"-m", invalidUrl});
+      try {
+        StaticLinkageCheckOption.configureMavenRepositories(commandLine);
+        Assert.fail("URL " + invalidUrl + " should be invalidated");
+      } catch (ParseException ex) {
+        // pass
+      }
     }
   }
 
@@ -112,7 +114,7 @@ public class StaticLinkageCheckOptionTest {
     CommandLine commandLine =
         StaticLinkageCheckOption.readCommandLine(new String[] {"-m", "file:///var/tmp"});
 
-    // This method should not raise exception
+    // This method should not raise an exception
     StaticLinkageCheckOption.configureMavenRepositories(commandLine);
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
@@ -101,7 +101,7 @@ public class StaticLinkageCheckOptionTest {
       throws ParseException, RepositoryException {
     CommandLine commandLine =
         StaticLinkageCheckOption.readCommandLine(
-            new String[] {"-m", "https://repo.spring.io/milestone"});
+            "-m", "https://repo.spring.io/milestone");
     StaticLinkageCheckOption.setRepositories(commandLine);
 
     // This artifact does not exist in Maven central, but it is in Spring's repository
@@ -117,7 +117,7 @@ public class StaticLinkageCheckOptionTest {
       throws ParseException {
     CommandLine commandLine =
         StaticLinkageCheckOption.readCommandLine(
-            new String[] {"-m", "https://repo.spring.io/milestone", "--no-maven-central"});
+            "-m", "https://repo.spring.io/milestone", "--no-maven-central");
     StaticLinkageCheckOption.setRepositories(commandLine);
 
     CollectRequest collectRequest = new CollectRequest();
@@ -140,7 +140,7 @@ public class StaticLinkageCheckOptionTest {
   private static void assertMavenRepositoryIsInvalid(String repositoryUrl)
       throws ParseException {
     CommandLine commandLine =
-        StaticLinkageCheckOption.readCommandLine(new String[]{"-m", repositoryUrl});
+        StaticLinkageCheckOption.readCommandLine("-m", repositoryUrl);
     try {
       StaticLinkageCheckOption.setRepositories(commandLine);
       Assert.fail("URL " + repositoryUrl + " should be invalidated");
@@ -152,7 +152,7 @@ public class StaticLinkageCheckOptionTest {
   @Test
   public void testConfigureAdditionalMavenRepositories_fileRepositoryUrl() throws ParseException {
     CommandLine commandLine =
-        StaticLinkageCheckOption.readCommandLine(new String[] {"-m", "file:///var/tmp"});
+        StaticLinkageCheckOption.readCommandLine("-m", "file:///var/tmp");
 
     // This method should not raise an exception
     StaticLinkageCheckOption.setRepositories(commandLine);
@@ -162,7 +162,7 @@ public class StaticLinkageCheckOptionTest {
   public void testReadCommandLine_multipleRepositoriesSeparatedByComma() throws ParseException {
     CommandLine commandLine =
         StaticLinkageCheckOption.readCommandLine(
-            new String[]{"-m", "file:///var/tmp,https://repo.spring.io/milestone"});
+            "-m", "file:///var/tmp,https://repo.spring.io/milestone");
     Truth.assertThat(commandLine.getOptionValues("m")).hasLength(2);
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
@@ -26,6 +26,8 @@ import org.apache.commons.cli.ParseException;
 import org.eclipse.aether.RepositoryException;
 import org.eclipse.aether.artifact.Artifact;
 import org.eclipse.aether.artifact.DefaultArtifact;
+import org.eclipse.aether.collection.CollectRequest;
+import org.eclipse.aether.repository.RemoteRepository;
 import org.junit.After;
 import org.junit.Assert;
 import org.junit.Test;
@@ -34,8 +36,8 @@ public class StaticLinkageCheckOptionTest {
 
   @After
   public void cleanup() {
-    // Resets the effect of configureMavenRepositories
-    RepositoryUtility.configureMavenRepositories(ImmutableList.of(), true);
+    // Resets the effect of setRepositories
+    RepositoryUtility.setRepositories(ImmutableList.of(), true);
   }
 
   @Test
@@ -118,14 +120,12 @@ public class StaticLinkageCheckOptionTest {
             new String[] {"-m", "https://repo.spring.io/milestone", "--no-maven-central"});
     StaticLinkageCheckOption.configureMavenRepositories(commandLine);
 
-    Artifact artifact = new DefaultArtifact("io.grpc:grpc-auth:1.15.1");
+    CollectRequest collectRequest = new CollectRequest();
+    RepositoryUtility.addRepositoriesToRequest(collectRequest);
 
-    try {
-      ClassPathBuilder.artifactsToClasspath(ImmutableList.of(artifact));
-      Assert.fail("gRPC artifact should not have been resolved without Maven Central");
-    } catch (RepositoryException ex) {
-      // pass
-    }
+    List<RemoteRepository> actualRepositories = collectRequest.getRepositories();
+    Truth.assertThat(actualRepositories).hasSize(1);
+    Truth.assertThat(actualRepositories).isEqualTo("repo.spring.io");
   }
 
   @Test

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckOptionTest.java
@@ -16,6 +16,7 @@
 
 package com.google.cloud.tools.opensource.classpath;
 
+import com.google.cloud.tools.opensource.dependencies.RepositoryUtility;
 import com.google.common.collect.ImmutableList;
 import com.google.common.truth.Truth;
 import java.nio.file.Path;
@@ -25,10 +26,17 @@ import org.apache.commons.cli.ParseException;
 import org.eclipse.aether.RepositoryException;
 import org.eclipse.aether.artifact.Artifact;
 import org.eclipse.aether.artifact.DefaultArtifact;
+import org.junit.After;
 import org.junit.Assert;
 import org.junit.Test;
 
 public class StaticLinkageCheckOptionTest {
+
+  @After
+  public void cleanup() {
+    // Resets the effect of configureMavenRepositories
+    RepositoryUtility.mavenRepositories = ImmutableList.of(RepositoryUtility.CENTRAL);
+  }
 
   @Test
   public void parseCommandLineOptions_shortOptions_bom() throws ParseException {
@@ -57,7 +65,7 @@ public class StaticLinkageCheckOptionTest {
 
   @Test
   public void parseCommandLineOptions_invalidOption() {
-    String[] arguments = {"-x"}; // No such option 
+    String[] arguments = {"-x"}; // No such option
     try {
       StaticLinkageCheckOption.readCommandLine(arguments);
       Assert.fail();
@@ -80,5 +88,31 @@ public class StaticLinkageCheckOptionTest {
 
     List<Path> paths = ClassPathBuilder.artifactsToClasspath(ImmutableList.of(artifact));
     Truth.assertThat(paths).isNotEmpty();
+  }
+
+  @Test
+  public void testConfigureAdditionalMavenRepositories_invalidRepositoryUrl()
+      throws ParseException {
+    CommandLine commandLine =
+        StaticLinkageCheckOption.readCommandLine(
+            new String[] {"-m", "https://repo.spring.io/milestone", "-m", "foobar"});
+
+    try {
+      StaticLinkageCheckOption.configureMavenRepositories(commandLine);
+      Assert.fail();
+    } catch (ParseException ex) {
+      // pass
+      Truth.assertThat(ex.getMessage())
+          .isEqualTo("Invalid URL specified for maven repository: foobar");
+    }
+  }
+
+  @Test
+  public void testConfigureAdditionalMavenRepositories_fileRepositoryUrl() throws ParseException {
+    CommandLine commandLine =
+        StaticLinkageCheckOption.readCommandLine(new String[] {"-m", "file:///var/tmp"});
+
+    // This method should not raise exception
+    StaticLinkageCheckOption.configureMavenRepositories(commandLine);
   }
 }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java
@@ -648,7 +648,7 @@ public class StaticLinkageCheckerTest {
     String bomCoordinates = "com.google.cloud:cloud-oss-bom:pom:1.0.0-SNAPSHOT";
 
     CommandLine parsedOption =
-        StaticLinkageCheckOption.readCommandLine(new String[] {"-b", bomCoordinates});
+        StaticLinkageCheckOption.readCommandLine("-b", bomCoordinates);
     ImmutableList<Path> inputClasspath =
         StaticLinkageCheckOption.generateInputClasspath(parsedOption);
     Truth.assertThat(inputClasspath).isNotEmpty();
@@ -702,7 +702,7 @@ public class StaticLinkageCheckerTest {
     // the unavailable dependency is under certain condition
     CommandLine parsedOption =
         StaticLinkageCheckOption.readCommandLine(
-            new String[] {"--artifacts", "com.google.guava:guava-gwt:20.0"});
+            "--artifacts", "com.google.guava:guava-gwt:20.0");
 
     ImmutableList<Path> inputClasspath =
         StaticLinkageCheckOption.generateInputClasspath(parsedOption);
@@ -720,7 +720,7 @@ public class StaticLinkageCheckerTest {
     //     org.eclipse.jdt.core.compiler:ecj:jar:4.4RC4 (not found in Maven central)
     CommandLine parsedOption =
         StaticLinkageCheckOption.readCommandLine(
-            new String[] {"--artifacts", "org.apache.tomcat:tomcat-jasper:8.0.9"});
+            "--artifacts", "org.apache.tomcat:tomcat-jasper:8.0.9");
 
     try {
       StaticLinkageCheckOption.generateInputClasspath(parsedOption);

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java
@@ -737,8 +737,8 @@ public class StaticLinkageCheckerTest {
   public void testGenerateInputClasspath_jarFileList()
       throws RepositoryException, ParseException {
 
-    String[] arguments = {"--jars", "dir1/foo.jar,dir2/bar.jar,baz.jar"};
-    CommandLine parsedOption = StaticLinkageCheckOption.readCommandLine(arguments);
+    CommandLine parsedOption =
+        StaticLinkageCheckOption.readCommandLine("--jars", "dir1/foo.jar,dir2/bar.jar,baz.jar");
     List<Path> inputClasspath = StaticLinkageCheckOption.generateInputClasspath(parsedOption);
 
     Truth.assertThat(inputClasspath)

